src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets

```diff
@@ -819,6 +819,8 @@ Copyright (c) .NET Foundation. All rights reserved.
 
   <Target Name="PublishItemsOutputGroup" DependsOnTargets="$(PublishItemsOutputGroupDependsOn)" Returns="@(PublishItemsOutputGroupOutputs)">
     <ItemGroup>
+      <PublishItemsOutputGroupOutputs Condition="'$(PublishDepsFilePath)' != '' and Exists($(PublishDepsFilePath))" Include="$(PublishDepsFilePath)" TargetPath="$(ProjectDepsFileName)" />
```

Comment:
Actually, the more I think about this, the more it doesn't matter to the output group whether it's the build or publish deps file.

What matters is if `GenerateDependencyFile` is `true`, then there will be a deps file in the publish output at `$(PublishDir)$(ProjectDepsFileName)`.  If we're using the build deps file, it will be copied to that location and already be an item in `ResolvedFileToPublish`.  If we're not using the build deps file, we generate a new one at that location.

What we really need here is this, a single entry to include the deps file if it isn't already in `ResolvedFileToPublish`:

` <PublishItemsOutputGroupOutputs Condition="'$(GenerateDependencyFile)' == 'true' and '$(_UseBuildDependencyFile)' != 'true'" Include="$(PublishDir)$(ProjectDepsFileName)" TargetPath="$(ProjectDepsFileName)" />
```

I think we can revert the other changes.  Sorry for my confusion.

Reply:
That doesn't work, since there's never any actual publish operation that creates the publish dir and it's contents.

I think my original change is correct, I've reverted the rest.  Don't think about this output group as the items that _were_ generated after a publish, think about it as the items that _would_ be collected by a publish operation.  In practice just using ProjectDepsFilePath works perfectly.

Reply:
If `GenerateDependencyFile` is `false` (say your project targets a .NET Framework TFM), no deps file will be generated, so I would expect the output group to reflect that.

Additionally, if `_UseBuildDependencyFile` is true, then I would expect `ResolvedFileToPublish` to [already contain the deps file](https://github.com/dotnet/sdk/blob/master/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets#L339-L343).  Are we sure that isn't the case?

If so, is there a specific problem with the suggested line above?

Reply:
The main problem with that line is that proposed file path:

Include="$(PublishDir)$(ProjectDepsFileName)"

The publish dir isn't created in this case, so that path doesn't work.  Using ProjectDepsFilePath works, as it's set to a path like "WpfApp4\WpfApp4\bin\x86\Debug\netcoreapp3.0\win-x86\WpfApp4.deps.json"

However, the condition part of your suggestion makes sense, I'll try it out.

Reply:
I'm confused what the use of the output group is then if all the files need to exist before the publish operation, since that would naturally exclude anything the publish operation generates.  Should it not describe the "output" of the publish operation were it run?

Reply:
With this change this output group contains the output of ComputeFilesToPublish + the deps.json file (when necessary).  The primary use case is VS installer projects, adding the contents of this output group to an installer project will give it all files necessary to run the app.  We can't actually run the entire publishing operation because installer projects have no knowledge of the publishing profile.  For this reason the self contained .net core 3 scenario requires specifying the RID directly in the project file (this is something we may look into improving later).

I added your suggested condition to handle the case where the deps file isn't needed.

Reply:
Ok, it just seems odd to me that it's a "publish" output group, but contains build output items.  They aren't necessarily the same thing as the publish operation itself might create files that do not exist as build outputs.  If I were a user, I'd soft of expect the installer to be operating off the publish output itself, but I can understand the limitation at work here.

Reply:
Actually, I'm not sure this is going to work as you have it.  If I build an application that has assets marked as excluded from publish, the build output currently will depend on those assets in the *build* deps file.  If you publish the application, those assets are excluded from `ResolvedFileToPublish`.  Copying the *build* deps file might not work because it'll have dependency information on assets that aren't being copied.

Can you check if this solution works with a dependency on a package that has `Publish="false"` in its `PackageReference` metadata?

Reply:
Regarding `Publish="false"`, I am going to propose making that exclude assets from build as well, so eventually the difference between the deps files might not matter here.

Reply:
I'm fully onboard with that and that would fix this particular deps file problem (assuming there is one).  It wouldn't resolve the issue where the publish target generates additional files, though.

Reply:
I see the publish=false issue you're talking about.  It's true that if this is set on a package reference my installer will have the dependency listed in the deps.json file but the actual assembly won't be there.  Are you fairly confident that this difference between the deps files will go away for 3.0?

Peter, I'm open to renaming this output group if you think PublishItems is misleading.

Reply:
We're fairly confident, I think.

I'm fine with taking this as-is to unblock your scenario, if it does so.  @dsplaisted do you have any additional feedback?

Reply:
@peterhuene - Thanks Peter, please sign off and merge this PR for me when you have a chance